### DISCUSSION TITLE
Woolley Ewelink CK-BL702-SWP-01(7020) - labelled as BSD29

### DISCUSSION THREAD

**Discussion by stain3565** [Koenkk/zigbee2mqtt#19473] [Devices]
Bought these at [Amazon](https://www.amazon.co.uk/dp/B0CDGKTTQP/ref=pe_27063361_487055811_TE_dp_1)

CK-BL702-SWP-01(7020) is not a recognised device in Z2M although there is a Tuya BSD29 device.

This is my first time at adding an unsupported device so not sure about the js required for this. However, I have a very basic js written that certainly turns the switch on and off.

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['CK-BL702-SWP-01(7020)'],
    model: 'BSD29',
    vendor: 'Ewelink',
    description: 'Woolley Zigbee plug',
    fromZigbee: [fz.on_off],
    toZigbee: [tz.on_off],
    exposes: [e.switch(),e.current(),e.power(),e.voltage()],
    // The configure method below is needed to make the device reports on/off state changes
    // when the device is controlled manually through the button on it.
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(1);
        await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff']);
        await reporting.onOff(endpoint);
    },
};

module.exports = definition;
```

So I at least know it is integrated. I am just not sure how the exposed elements are:

a. Identified
b. Populated

I have added current, power and voltage but these are showing as N/A.

I realise that these plugs may require polling to get the info but I haven't found any examples explaining what the js is for this. Also, as stated, I have not sussed how I identify the exposed attributes.

Some debug output is below although I don't believe that gives any useful info:

`Debug 2023-10-29 13:08:16Received Zigbee message from 'Woolley Plug 1', type 'read', cluster 'genTime', data '["localTime"]' from endpoint 1 with groupID 0
Debug 2023-10-29 13:08:23Received Zigbee message from 'Woolley Plug 1', type 'read', cluster 'genTime', data '["localTime"]' from endpoint 1 with groupID 0
Debug 2023-10-29 13:08:30Received Zigbee message from 'Woolley Plug 1', type 'read', cluster 'genTime', data '["localTime"]' from endpoint 1 with groupID 0
Debug 2023-10-29 13:08:37Received Zigbee message from 'Woolley Plug 1', type 'read', cluster 'genTime', data '["localTime"]' from endpoint 1 with groupID 0
Debug 2023-10-29 13:08:44Received Zigbee message from 'Woolley Plug 1', type 'read', cluster 'genTime', data '["localTime"]' from endpoint 1 with groupID 0
Debug 2023-10-29 13:08:51Received Zigbee message from 'Woolley Plug 1', type 'read', cluster 'genTime', data '["localTime"]' from endpoint 1 with groupID 0
Debug 2023-10-29 13:08:55Received Zigbee message from 'Woolley Plug 1', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 1 with groupID 0
Info 2023-10-29 13:08:55MQTT publish: topic 'zigbee2mqtt/Woolley Plug 1', payload '{"current":null,"linkquality":87,"power":null,"state":"ON","voltage":null}'
Debug 2023-10-29 13:08:55Received Zigbee message from 'Woolley Plug 1', type 'attributeReport', cluster 'genOnOff', data '{"onOff":1}' from endpoint 1 with groupID 0
Info 2023-10-29 13:08:55MQTT publish: topic 'zigbee2mqtt/Woolley Plug 1', payload '{"current":null,"linkquality":90,"power":null,"state":"ON","voltage":null}'
Debug 2023-10-29 13:08:57Received Zigbee message from 'Woolley Plug 1', type 'attributeReport', cluster 'genOnOff', data '{"onOff":0}' from endpoint 1 with groupID 0
Info 2023-10-29 13:08:57MQTT publish: topic 'zigbee2mqtt/Woolley Plug 1', payload '{"current":null,"linkquality":87,"power":null,"state":"OFF","voltage":null}'
Debug 2023-10-29 13:08:57Received Zigbee message from 'Woolley Plug 1', type 'attributeReport', cluster 'genOnOff', data '{"onOff":0}' from endpoint 1 with groupID 0
Info 2023-10-29 13:08:57MQTT publish: topic 'zigbee2mqtt/Woolley Plug 1', payload '{"current":null,"linkquality":84,"power":null,"state":"OFF","voltage":null}'
Debug 2023-10-29 13:09:05Received Zigbee message from 'Woolley Plug 1', type 'read', cluster 'genTime', data '["localTime"]' from endpoint 1 with groupID 0
`

Any ideas what js should support this?

**Comment by stain3565** [comment]
I have copied the created database entry below

`{
	"id":50,
	"type":"Router",
	"ieeeAddr":"0x7cb94c76a6740000",
	"nwkAddr":10780,
	"manufId":4742,
	"manufName":"eWeLink",
	"powerSource":"Unknown",
	"modelId":"CK-BL702-SWP-01(7020)",
	"epList":[1,242],
	"endpoints":{"1":
					{"profId":260,
					"epId":1,
					"devId":9,
					"inClusterList":[0,3,4,5,6,4096,64599,64529],
					"outClusterList":[25,2820,10],
					"clusters":
						{"genBasic":
							{"attributes":
								{"stackVersion":0,"hwVersion":0}
							},
						"genOnOff":
							{"attributes":
								{"onOff":0}
							}
						},
					"binds":
						[{"cluster":6,
						"type":"endpoint",
						"deviceIeeeAddress":"0x00124b0024c943f4",
						"endpointID":1}],
					"configuredReportings":
						[{"cluster":6,
						"attrId":0,
						"minRepIntval":0,
						"maxRepIntval":3600,
						"repChange":0,
						"manufacturerCode":null}],
					"meta":{}
					},
				"242":
					{"profId":41440,
					"epId":242,
					"devId":97,
					"inClusterList":[],
					"outClusterList":[33],
					"clusters":{},"binds":[],
					"configuredReportings":[],"meta":{}
					}
				},
	"appVersion":1,
	"stackVersion":0,
	"hwVersion":0,
	"dateCode":"20210610-BL",
	"swBuildId":"1.0.1",
	"zclVersion":8,
	"interviewCompleted":true,
	"meta":{"configured":1466202208},
	"lastSeen":1698590440936,
	"defaultSendRequestWhen":"immediate"
}
`

**Comment by Koenkk** [comment]
Duplicate of https://github.com/Koenkk/zigbee2mqtt/issues/19334
